[1.19.3] Replace AdvancementProvider patch with Forge helper

ChampionAsh5357 opened the pull request:
Vanilla data generators have been abstracted such that implementations only need to implement a single method or pass in sub providers into the constructor. This works for most providers except for the `AdvancementProvider`. This is because to generate an advancement with an existing parent, the `ExistingFileHelper` is needed along with a patch to the `Advancement$Builder` to check the parent without always crashing. As such, the patch to `AdvancementProvider` can be removed without issue.

Normally, this could be it as the `ExistingFileHelper` could just be passed into each sub provider. However, this is not as modder-friendly as many would just be inclined to use the regular build methods which, as mentioned, crash. Therefore, a provider is added that extends `AdvancementProvider` and provides generators which take in the `ExistingFileHelper`.

I believe this solution is the best, but it's not backed by anything as this is not the only solution; there could just be a `Function<ExistingFileHelper, AdvancementSubProvider>` list or just disregarding that entirely.

ChampionAsh5357 committed:
Remove patches and add helper instead

sciwhiz12 commented on src/main/java/net/minecraftforge/common/data/ForgeAdvancementProvider.java:
```diff
@@ -0,0 +1,62 @@
+package net.minecraftforge.common.data;
```
Missing license header, please run the `updateLicenses`/`licenseFormat` task.

ChampionAsh5357 replied:
You know, I thought I did this, but that might have been a time before I deleted and recreated the file a bunch.

sciwhiz12 commented on src/main/java/net/minecraftforge/common/data/ForgeAdvancementProvider.java:
```diff
@@ -0,0 +1,62 @@
+ * An extension of the {@link AdvancementProvider} to provide a feature-complete
+ * experience to generate modded advancements.
+ */
+public abstract class ForgeAdvancementProvider extends AdvancementProvider
```
There is the issue of discoverability for this class, in two parts: modders are unlikely to discover this class as `AdvancementProvider` provides no hints as to its existence (other than perhaps the IDE's hint that there is another subclass), and that all other classes with the `Forge` prefix in the package are used by Forge itself and not meant for subclassing (which means modders may infer this class too should not be used). 

Should we resolve this somehow, in-code through comments or otherwise through announcements/docs, or let modders figure this out themselves?

ChampionAsh5357 replied:
I'm just gonna add a comment with no deprecation to the provider itself. It's better to have backtraces for understanding.

sciwhiz12 commented on src/main/java/net/minecraftforge/common/data/ForgeAdvancementProvider.java:
```diff
@@ -0,0 +1,62 @@
+     * @param existingFileHelper a helper used to find whether a file exists
+     * @param subProviders the generators used to create the advancements
+     */
+    public ForgeAdvancementProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> registries, ExistingFileHelper existingFileHelper, List<AdvancementGenerator> subProviders)
```
Just noting that Mojang calls the future as `lookupProvider`. (For reference, see the fields in `TagsProvider`.) It may preferrable to rename to that, for consistency, but I leave that decision to you.

ChampionAsh5357 replied:
Mojang also calls it `registries` in the advancement provider so I'm gonna stick with that.

sciwhiz12 commented on patches/minecraft/net/minecraft/data/advancements/AdvancementProvider.java.patch:
```diff
@@ -1,53 +0,0 @@
-+      this(p_256529_, p_255722_, p_255883_, null);
-+   }
-+
-+   public AdvancementProvider(PackOutput p_256529_, CompletableFuture<HolderLookup.Provider> p_255722_, List<AdvancementSubProvider> p_255883_, @org.jetbrains.annotations.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
```
Please remove the corresponding EXC entry:

https://github.com/MinecraftForge/MinecraftForge/blob/ec3abdea1455fc8c4e0f440d77e2dc24f880a4a6/src/main/resources/forge.exc#L34

Just to be clear, the `checkExc` task did find this but even though we run it in our CI, it does not error out on failure; instead, [it logged the line and silently corrected the problem](https://teamcity.minecraftforge.net/buildConfiguration/MinecraftForge_MinecraftForge_MinecraftForge_MinecraftForge__PullRequestChecks/6004?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&showLog=6004_769_769&expandPull+Request+Details=true&expandBuildProblemsSection=true&logView=flowAware) in the CI agent's local copy.

ChampionAsh5357 replied:
This should really output a warning then imo, silent checks are not great especially if cleanup is happening.

sciwhiz12 replied:
Yeah, that's something which we need to work on with the check tasks in general.

sciwhiz12 replied:
Just to note, I've filed an issue about this: #9207.

ChampionAsh5357 committed:
Address comments

- Remove exc entry
- Add license to Forge provider
- Add comment directing modders to use ForgeAdvancementProvider instead

LexManos commented on patches/minecraft/net/minecraft/data/advancements/AdvancementProvider.java.patch:
```diff
@@ -1,53 +1,10 @@
+ import net.minecraft.resources.ResourceLocation;
+ 
++// Forge: Use ForgeAdvancementProvider instead, provides ease of access for the ExistingFileHelper in the generator
+ public class AdvancementProvider implements DataProvider {
```
Shouldnt this have  proper javadoc comment with deprecation instead of just a single line comment?

ChampionAsh5357 replied:
I thought about it, but there is nothing wrong with using the original advancement provider and passing in the existing file helper yourself. Though, if we want to discourage that, I can update it.

LexManos replied:
If we push people twards a new version, we tend to deprecate.

ChampionAsh5357 replied:
Ok, changing it now then.

ChampionAsh5357 replied:
Done

ChampionAsh5357 committed:
Deprecate and link to Forge provider
